motif/matchComp.go

```diff
@@ -0,0 +1,288 @@
+package motif
+
+import (
+	"fmt"
+	"github.com/vertgenlab/gonomics/bed"
+	"github.com/vertgenlab/gonomics/dna"
+	"github.com/vertgenlab/gonomics/exception"
+	"github.com/vertgenlab/gonomics/fasta"
+	"github.com/vertgenlab/gonomics/fileio"
+	"github.com/vertgenlab/gonomics/numbers"
+	"log"
+	"math"
+)
+
+type MatchCompSettings struct {
+	MotifFile          string
+	MotifType          string
+	Records            []fasta.Fasta
+	PropMatch          float64
+	ChromName          string
+	OutFile            string
+	Pseudocounts       float64
+	ResidualWindowSize int
+	RefStart           int
+	OutputAsProportion bool
+	EnforceStrandMatch bool
+	ResidualFilter     float64
+}
+
+func MatchComp(s MatchCompSettings) {
+	var err error
+	var motifLen int
+	var kmerHash, revKmerHash map[uint64]float64
+	var consensusScore float64
+	var couldScoreConsensus bool
+	var revCompMotif PositionMatrix
+
+	var motifs []PositionMatrix
+	switch s.MotifType {
+	case "Frequency":
+		motifs = ReadJaspar(s.MotifFile, "Frequency")
+		motifs = PfmSliceToPpmSlice(motifs, s.Pseudocounts)
+		motifs = PpmSliceToPwmSlice(motifs)
+	case "Probability":
+		motifs = ReadJaspar(s.MotifFile, "Probability")
+		motifs = PpmSliceToPwmSlice(motifs)
+	case "Weight":
+		motifs = ReadJaspar(s.MotifFile, "Weight")
+	default:
+		log.Fatalf("Error. Unexpected motif file format. Options are 'Frequency', 'Probability', and 'Weight'.")
+	}
+
+	out := fileio.EasyCreate(s.OutFile)
+
+	for i := range motifs {
+		motifLen = len(motifs[i].Mat[0])
+		if motifLen > 32 {
+			log.Fatalf("Error: MatchComp cannot accommodate Position Matrices with a motif length greater than 32. Plese filter out the matrix with this ID: %v.\n", motifs[i].Id)
+		}
+		var currSeq = ConsensusSequence(motifs[i], false)
+		consensusScore, _, couldScoreConsensus = ScoreWindow(motifs[i], currSeq.Seq, 0)
+		if !couldScoreConsensus {
+			log.Fatalf("Error: problem in buildKmerHash. Could not score consensus sequence.")
+		}
+
+		altEndsConsidered := make(map[int]bool, 0)
+
+		kmerHash = buildKmerHash(motifs[i], s.PropMatch)
+		scanRefSequenceComp(s.Records, kmerHash, motifs[i], s.ChromName, out, s.ResidualWindowSize, consensusScore, bed.Positive, s.RefStart, s.EnforceStrandMatch, s.OutputAsProportion, altEndsConsidered, s.ResidualFilter)
+
+		revCompMotif = ReverseComplement(motifs[i])
+		revKmerHash = buildKmerHash(revCompMotif, s.PropMatch)
+		scanRefSequenceComp(s.Records, revKmerHash, revCompMotif, s.ChromName, out, s.ResidualWindowSize, consensusScore, bed.Negative, s.RefStart, s.EnforceStrandMatch, s.OutputAsProportion, altEndsConsidered, s.ResidualFilter)
+
+		//now we scan the alt sequences for any motifs lost in ref
```

Comment:
is "alt" the same as "query", aka the 2nd sequence in the multiFa (e.g. chimp), as opposed to the reference/target /1st sequence in the multiFa (e.g. human)?

Reply:
I apologize if I have used this inconsistently. Should we standardize with 'target' 'query'?

Reply:
No worries! 'target' 'query' seems more intuitive and generalized to me, because when I see 'ref' 'alt' I think of segregating sites in vcfs so they give me the feeling that they can only describe 2 versions/populations of the same species, not 2 species, but tfMatchComp can be used to compare any 2 fastas including 2 different species.